ApiDoctor.Publishing/CSDL/csdlwriter.cs

```diff
@@ -1851,8 +1844,164 @@ private static void AddReadByKeyRestriction(IODataAnnotatable annotatable, Metho
+        {
+           { "GET", Term.LinkRel.List },
+           { "POST", Term.LinkRel.Create },
+           { "PATCH", Term.LinkRel.Update },
+           { "PUT", Term.LinkRel.Update },
+           { "DELETE", Term.LinkRel.Delete }
+        };
+
+        private static string GetLinkRelValueForMethod(string httpMethod)
+        {
+            httpMethodToLinkRelMappings.TryGetValue(httpMethod, out string linkRel);
+            return linkRel;
+        }
+
+        private static void AddLinkAnnotation(IODataAnnotatable annotatable, MethodDefinition sourceMethod)
+        {
+            var linkRel = GetLinkRelValueForMethod(sourceMethod.HttpMethodVerb());
+            if (linkRel == null) return;
+
+            var linkRecord = CreateLinksRecord(sourceMethod.SourceFile.DisplayName, linkRel);
+
+            var linkAnnotation = annotatable.Annotation.FirstOrDefault(x => x.Term == Term.LinksTerm);
+            if (linkAnnotation == null)
+            {
+                linkAnnotation = new Annotation
+                {
+                    Term = Term.LinksTerm,
+                    Collection = new RecordCollection
+                    {
+                        Records = new List<Record> { linkRecord }
+                    }
+                };
+                annotatable.Annotation.Add(linkAnnotation);
+            }
+            else
+            {
+                var linkRecordExists = linkAnnotation.Collection.Records
+                   .Where(record => record.Type == linkRecord.Type)
+                   .SelectMany(p => p.PropertyValues)
+                   .Any(d => (d.Property == linkRecord.PropertyValues[0].Property && d.String == linkRecord.PropertyValues[0].String));
+                if (!linkRecordExists)
+                {
+                    linkAnnotation.Collection.Records.Add(linkRecord);
+                }
+            }
+        }
+
+        private static void AddReadByKeyLinkAnnotation(IODataAnnotatable annotatable, MethodDefinition sourceMethod)
+        {
+            var readRestriction = annotatable.Annotation
+                .FirstOrDefault(annotation => annotation.Term == Term.LinksTerm);
+            var currentLinkRecord = CreateLinksRecord(sourceMethod.SourceFile.DisplayName, Term.LinkRel.ReadByKey);
+            if (readRestriction != null)
+            {
+                var propertyValue = readRestriction.Collection.Records
+                    .SelectMany(c => c.PropertyValues)
+                    .FirstOrDefault(x => x.String == Term.LinkRel.ReadByKey);
+                if (propertyValue == null)
+                {
+                    readRestriction.Collection.Records.Add(currentLinkRecord);
+                }
+            }
+            else
+            {
+                AddLinkAnnotation(annotatable, sourceMethod, Term.LinksTerm, currentLinkRecord);
+                AddReadByKeyLinkAnnotation(annotatable, sourceMethod);
```

Comment:
Since this function is recursive, it looks like there is a case where the recursion may never terminate if `AddLinkAnnotation` does not add the link and exits through line 1917. We should probably verify the link is added (maybe a boolean return type for `AddLinkAnnotation`?) before making the recursive call to avoid this scenario.